### ISSUE TITLE
Inter-robot communication

### ISSUE THREAD

**Issue by byorgey** [byorgey/swarm#94]
This may be still a ways off, but eventually we certainly want to be able to have inter-robot communication.  This would enable lots of cool things such as remotely controlling or reprogramming robots, creating teams of robots that can dynamically coordinate among one another, etc. There is some discussion of this in https://github.com/byorgey/swarm/issues/17#issuecomment-927669308 .

There is a lot of language design work to be done here.  My initial vague thought to look for inspiration to things like pi-calculus and session types.  Probably full-on session types are too heavy-handed, though.  Maybe something simple like typed broadcast channels: a value of type `chan t` is a channel over which values of type `t` can be broadcast, supporting operations like
```
newChannel : cmd (chan t)
broadcast : forall t. chan t -> t -> cmd ()
read : forall t. chan t -> cmd t
readNonBlocking : forall t. chan t -> cmd (maybe t)
```
Channels will be private in the sense that only robots with a reference to a channel can use it.  There would have to be some special support for `newChannel` similar to `def`, so a robot can declare a new channel and then have it available globally to all robots it built after that.

There are also game design issues to think about; e.g. what sort(s) of device(s) should be needed to use these features?

**Comment by byorgey** [comment]
I'm upgrading this from `Nice to have` to `Moderate` --- it seems to keep coming up!  I think this is probably one of the first things we should tackle after the alpha release.

**Comment by byorgey** [comment]
We should also put a bit of thought into propagation delay: the time it takes for a `broadcast` message to reach a listener can be proportional to the distance.

**Comment by xsebek** [comment]
> I'm upgrading this from Nice to have to Moderate --- it seems to keep coming up! I think this is probably one of the first things we should tackle after the alpha release.

@byorgey shall I pin this then? :slightly_smiling_face:

**Comment by xsebek** [comment]
> We should also put a bit of thought into propagation delay: the time it takes for a broadcast message to reach a listener can be proportional to the distance.

I just realized that if you send a robot to walk away, then we would have to store all messages that the robot could possibly have access to by the time they reach it. That could easily be more than 10000 if you have a lot of robots broadcasting a lot of messages.

**Comment by byorgey** [comment]
Yeah, that gets complicated.  For the sake of argument, what if we just limit the broadcast radius but have instantaneous propagation within that radius?  And for I/O cables what if we just allow instantaneous transmission for any distance?  That is "un-physical", of course, but is there any way it would mess up the game or make things inconsistent?

**Comment by xsebek** [comment]
Well, if the signal propagation was by tiles, then it might be manageable.

We could discard the message once it reaches the most distant known tile. If we ignore teleport, we can even precompute when will it _definitely_ reach that tile, because the most distant robot can only move so far after you send the message.

I am all for cables being instantaneous, that will at least keep them useful. :+1:
